src/ratingservice/main.py

```diff
@@ -13,103 +13,155 @@
+        return makeError(400, "missing payload")
+    return getRatingById(data['id'])
+
+
+@app.route('/rating', methods=['POST'])
+def postRating():
+    data = request.get_json()
+    if data == None:
+        return makeError(400, "missing payload")
+
+    eid = data['id']
+    if eid == None or eid == "":
+        return makeError(400, "malformed entity id")
+    rating_str = data['rating']
+    if rating_str == None or rating_str == "":
+        return makeError(400, "malformed rating")
+    try:
+        rating = int(rating_str)
+        if rating < 1 or rating > 5:
+            return makeError(400, "rating should be from 1 to 5")
+    except ValueError:
+        return makeError(400, "malformed rating")
+
+    conn = getConnection()
+    if conn == None:
+        return makeError(500, 'failed to connect to DB')
     try:
         with conn.cursor() as cursor:
-            cursor.execute("SELECT AVG(score), COUNT(*) FROM ratings WHERE product_id='{}';".format(id))
-            result = cursor.fetchone()
-            rating, count = result[0], result[1]
-            if count > 0:
-                # product exists
-                resp = jsonify({
-                    'status' : 'success',
-                    'rating' : str(rating),
-                    'count'  : str(count)
-                })
-                resp.status_code = 200
-            else:
-                # product not exists
-                resp = jsonify({'error' : 'Product not found'.format(id)})
-                resp.status_code = 404   
-        conn.commit()    
+            cursor.execute("INSERT INTO votes (eid, rating) VALUES ('{0}', {1});".format(
+                eid, rating))
+        conn.commit()
+        return makeResult({})
+    except IntegrityError:
+        return makeError(404, 'invalid entity id')
     except:
-        resp = jsonify({'error' : 'Error in database: Fail to get the rating of product {}'.format(id)})
-        resp.status_code = 500
+        return makeError(500, 'DB error')
     finally:
-        connpool.putconn(conn)
-    return resp
+        db_connection_pool.putconn(conn)
 
-# rate a product
-@app.route('/rating', methods=['POST'])
-def rate():
-    conn = connpool.getconn()
-    resp = None
-    product_id = request.form['id']
-    score = request.form['score']
+
+@app.route('/recollect', methods=['GET'])
+def aggregateRatings():
+    conn = getConnection()
+    if conn == None:
+        return makeError(500, 'failed to connect to DB')
     try:
         with conn.cursor() as cursor:
-            cursor.execute("INSERT INTO ratings (product_id, score) VALUES ('{0}', {1});".format(product_id, score))
-            resp = jsonify({'status' : 'success'})
-            resp.status_code = 200
+            cursor.execute("UPDATE votes SET in_process=TRUE;")
+            cursor.execute(
+                "UPDATE ratings AS r SET "
+                "rating=(r.rating*r.votes/(r.votes+v.votes))+(v.avg_rating*v.votes/(r.votes+v.votes)), "
```

Comment:
What are you doing here?

Reply:
+1, this could use some commenting at very least